### PR TITLE
Boost: Add cloud critical CSS generation feature (take 2)

### PR DESCRIPTION
**Pull request by thingalon**
Fixes 45-gh-Automattic/boost-hydra

#### Changes proposed in this Pull Request:
* Add a cloud CSS module based on critical CSS feature
* Make it possible to switch between cloud CSS and critical CSS, by defining `JETPACK_BOOST_CLOUD_CSS` as a boolean value.
* Refactor Critical CSS feature to move common classes to lib for sharing between the two modules.
* Add rest API endpoints to request and update cloud css

#### Jetpack product discussion
pc9hqz-fv-p2

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* Start the docker development environment `jetpack docker up`.
* Add a tunnel `jetpack docker jt-up` and access the site through the tunnel.
* Add the development MU plugin for testing local setup (shared internally). 
* Start shield in development mode `npm run docker:dev`
* Start hydra in connected mode `npm run connected-docker`
* Try enabling or requesting regenerate Cloud CSS.
* You should see critical CSS in the front-end source.

### CODE REVIEW

File: projects/plugins/boost/app/contracts/Boost_API_Client.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace Automattic\Jetpack_Boost\Contracts;
+
+/**
+ * An interface to build Boost API client.
+ *
+ * Communication with Boost back-end should be done through this interface.
+ */
+interface Boost_API_Client {
```
Comment (thingalon) [inline projects/plugins/boost/app/contracts/Boost_API_Client.php:10]: 👍

File: projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	private $cloud_css_state;
+
+	/**
+	 * @var Source_Providers
+	 */
+	private $source_providers;
+
+	public function __construct() {
+		$this->source_providers = new Source_Providers();
+		$this->cloud_css_state  = new Critical_CSS_State();
```
Comment (thingalon) [inline projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php:33]: Is it safe to create multiple of these as needed? Each one will read and write (and store) a copy of the transient separately.

Wouldn't it be safer to pass in a Critical_CSS_State object?

File: projects/plugins/boost/app/rest-api/endpoints/Request_Cloud_CSS.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<?php
+/**
+ * Create a new request for cloud critical CSS.
```
Comment (thingalon) [inline projects/plugins/boost/app/rest-api/endpoints/Request_Cloud_CSS.php:3]: Can we include the HTTP method and path in the endpoint comment?

**Review by thingalon — COMMENTED** [review]
👏 There is a lot to like here, and some very nicely written code.

A couple of comments / concerns inline.

File: projects/plugins/boost/app/rest-api/endpoints/Request_Cloud_CSS.php (hunk 3 above)
Comment (eAdnan007) [reply projects/plugins/boost/app/rest-api/endpoints/Request_Cloud_CSS.php:3]: Added

File: projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php (hunk 2 above)
Comment (eAdnan007) [reply projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS_Request.php]: I have refactored the `Critical_CSS_State` a bit. It is now possible to instantiate different states by naming them differently. You can also pass an instance to the request class.

**Review by eAdnan007 — APPROVED** [review]
👍🏼  on behalf of @thingalon

**Pull request merged**